### PR TITLE
Liquity updates

### PR DESCRIPTION
**Pull request by KaymasJain**
Updated the Liquity connector for proper composability with other connectors and covering some edge case of max amounts. Not updated the test cases yet.

### CODE REVIEW

File: contracts/mainnet/connectors/liquity/main.sol [hunk 1]
```diff
@@ -50,23 +29,23 @@ abstract contract LiquityResolver is Events, Helpers {
...
+        uint[] getIds,
+        uint[] setIds
     ) external payable returns (string memory _eventName, bytes memory _eventParam) {
-        if (getId != 0 && depositAmount != 0) {
-            revert("open(): Cannot supply a depositAmount if a non-zero getId is supplied");
-        }
 
-        depositAmount = getUint(getId, depositAmount);
+        depositAmount = getUint(getIds[0], depositAmount);
+        borrowAmount = getUint(getIds[1], borrowAmount);
```
Comment (edmulraney) [inline contracts/mainnet/connectors/liquity/main.sol:46]: Is the purpose of making the `borrowAmount` gettable to enable other spells to control amount that gets borrowed? What would some examples be where the `borrowAmount` needs to be gettable?

File: contracts/mainnet/connectors/liquity/main.sol [hunk 2]
```diff
@@ -75,25 +54,26 @@ abstract contract LiquityResolver is Events, Helpers {
             lowerHint
         );
 
-        // Allow other spells to use the borrowed amount
-        setUint(setId, borrowAmount);
-        _eventName = "LogOpen(address,uint256,uint256,uint256,uint256,uint256)";
-        _eventParam = abi.encode(msg.sender, maxFeePercentage, depositAmount, borrowAmount, getId, setId);
+        setUint(setIds[0], depositAmount);
```
Comment (edmulraney) [inline contracts/mainnet/connectors/liquity/main.sol:57]: ~~What would the use case be for setting a storage slot for the amount that was deposited?~~
Edit: I see this handles the case whereby the user supplies the max available balance `uint(-1)` and then this `setId` will give you the actual amount deposited. Nice! 
What would a use case of this be? Why might a subsequent spell need the amount deposited?

File: contracts/mainnet/connectors/liquity/main.sol [hunk 3]
```diff
@@ -50,23 +29,23 @@ abstract contract LiquityResolver is Events, Helpers {
...
         address upperHint,
         address lowerHint,
-        uint getId,
-        uint setId
+        uint[] getIds,
+        uint[] setIds
     ) external payable returns (string memory _eventName, bytes memory _eventParam) {
-        if (getId != 0 && depositAmount != 0) {
-            revert("open(): Cannot supply a depositAmount if a non-zero getId is supplied");
-        }
```
Comment (edmulraney) [inline contracts/mainnet/connectors/liquity/main.sol:67]: Is there a problem with this type of check?

File: contracts/mainnet/connectors/liquity/main.sol [hunk 4]
```diff
@@ -196,37 +200,40 @@ abstract contract LiquityResolver is Events, Helpers {
...
-        adjustTrove.borrowAmount = borrowAmount;
-        adjustTrove.repayAmount = getUint(getRepayId, repayAmount);
+
+        depositAmount = getUint(getIds[0], depositAmount);
+        adjustTrove.depositAmount = depositAmount == uint(-1) ? address(this).balance : depositAmount;
+
+        withdrawAmount = getUint(getIds[1], withdrawAmount);
+        adjustTrove.withdrawAmount = withdrawAmount == uint(-1) ? troveManager.getTroveColl(address(this)) : withdrawAmount;
+
+        adjustTrove.borrowAmount = getUint(getIds[2], borrowAmount);
```
Comment (edmulraney) [inline contracts/mainnet/connectors/liquity/main.sol:227]: Do we need to check if the `borrowAmount` is `uint(-1)` here also? So we can use the DSA's LUSD balance

File: contracts/mainnet/connectors/liquity/main.sol [hunk 5]
```diff
@@ -238,28 +245,27 @@ abstract contract LiquityResolver is Events, Helpers {
             lowerHint
         );
         
-        // Allow other spells to use the withdrawn collateral
-        setUint(setWithdrawId, withdrawAmount);
+        setUint(setIds[0], depositAmount);
+        setUint(setIds[1], withdrawAmount);
+        setUint(setIds[2], borrowAmount);
+        setUint(setIds[3], repayAmount);
```
Comment (edmulraney) [inline contracts/mainnet/connectors/liquity/main.sol]: Should these setId values use the final value contained inside `adjustTrove.*` struct? Otherwise they might not be the same values that were sent to Liquity

File: contracts/mainnet/connectors/liquity/main.sol [hunk 6]
```diff
@@ -238,28 +245,27 @@ abstract contract LiquityResolver is Events, Helpers {
...
 
-        // Allow other spells to use the borrowed amount
-        setUint(setBorrowId, borrowAmount);
-
-        _eventName = "LogAdjust(address,uint256,uint256,uint256,uint256,uint256,uint256,uint256,uint256,uint256)";
-        _eventParam = abi.encode(msg.sender, maxFeePercentage, depositAmount, withdrawAmount, borrowAmount, repayAmount, getDepositId, setWithdrawId, getRepayId, setBorrowId);
+        _eventName = "LogAdjust(address,uint256,uint256,uint256,uint256,uint256,uint256[],uint256[])";
+        _eventParam = abi.encode(address(this), maxFeePercentage, depositAmount, withdrawAmount, borrowAmount, repayAmount, getIds, setIds);
```
Comment (edmulraney) [inline contracts/mainnet/connectors/liquity/main.sol]: Same here, feels like they should be using `adjustTrove.*` values

File: contracts/mainnet/connectors/liquity/main.sol (hunk 1 above)
Comment (KaymasJain) [reply contracts/mainnet/connectors/liquity/main.sol:46]: In refinancing where you need to do 100% refinancing. Eg:- let's say LUSD is supported by Aave now their could be Aave & Liquity refinancing. When a user wants to move 100% debt, he'll borrow flashloan & payback 100% on Aave using `uint(-1)` and store it in an ID then use that ID to borrow the same amount from Liquity.

Also, in general, we prefer to keep things open even if we can't think of any use cases because it has happened with us before where we found a new use case and then have to deploy newly updated connector for it.

File: contracts/mainnet/connectors/liquity/main.sol (hunk 2 above)
Comment (KaymasJain) [reply contracts/mainnet/connectors/liquity/main.sol:57]: It might need is some heavier calculation. Let's say a user took a flashloan to deposit and now he wants to withdraw the same amount of assets from any other protocol to return to the flashloan. If the amount is sent from front end then there's no issue but the amount is calculating on the smart contract level then it matters when it's uncertain.

Also, as I mentioned above we prefer to keep things open because we don't know what other use cases could possibly be there in future.

File: contracts/mainnet/connectors/liquity/main.sol (hunk 3 above)
Comment (KaymasJain) [reply contracts/mainnet/connectors/liquity/main.sol:67]: I think it doesn't matter much. `getId` will overwrite the `depositAmount` if it's anything other than 0. So the checks aren't important here but the checks were not wrong.

File: contracts/mainnet/connectors/liquity/main.sol (hunk 4 above)
Comment (KaymasJain) [reply contracts/mainnet/connectors/liquity/main.sol:227]: How will we check `borrowAmount` as `uint(-1)` as borrow depends on user's position and not on user's LUSD balance. Basically, borrowing max can make position very risky and near to liquidation and will also add unnecessary logics.

File: contracts/mainnet/connectors/liquity/main.sol (hunk 5 above)
Comment (KaymasJain) [reply contracts/mainnet/connectors/liquity/main.sol]: Ooh! Sorry. My bad. We can update this.

Comment (KaymasJain) [reply contracts/mainnet/connectors/liquity/main.sol]: Done https://github.com/liquity/dsa-connectors/pull/2/commits/027095603ce7b8e37e0d31f0660834bfd0a91c86

File: contracts/mainnet/connectors/liquity/main.sol (hunk 6 above)
Comment (KaymasJain) [reply contracts/mainnet/connectors/liquity/main.sol]: Done https://github.com/liquity/dsa-connectors/pull/2/commits/d45f08b7430350a24555b2cd96b5834c72fd84c0

File: contracts/mainnet/connectors/liquity/main.sol (hunk 3 above)
Comment (edmulraney) [reply contracts/mainnet/connectors/liquity/main.sol:67]: Ah okay, I was trying to protect developers from incorrectly using the API since there’s no indicator that the value is ignored when a getId is supplied. This way there less likely to encounter bugs.

**Review by edmulraney — APPROVED** [review]
LGTM, thanks @KaymasJain for the PR :)

Before we merge it I will fix the broken tests

File: contracts/mainnet/connectors/liquity/main.sol (hunk 3 above)
Comment (KaymasJain) [reply contracts/mainnet/connectors/liquity/main.sol:67]: Yeah! I understood it. Although, the assets will either be in user's Liquity position or in DSA so it's alright. Anyway, we might think of doing these minor updates if makes sense in all our connectors at once but let's go like this for now :)
